src/msemu.c

```diff
@@ -809,54 +700,49 @@ int main(int argc, char *argv[])
 	 * If the dataflash image does not exist, it will be created when
 	 * the dataflash is written to disk.
 	 */
-	if (!flashtobuf((uint8_t *)ms.dev_map[DF], dataflash_path, MEBIBYTE/2)) {
-		printf("Existing dataflash image not found at '%s'.\n", dataflash_path);
+	if (!flashtobuf((uint8_t *)ms->dev_map[DF], options->df_path, MEBIBYTE/2)) {
+		printf("Existing dataflash image not found at '%s'.\n", options->df_path);
 	}
-	printf("Dataflash will be saved to '%s' on exit.\n", dataflash_path);
-
 
-	/* TODO: Add git tags to this, because thats neat */
-	printf("\nMailstation Emulator v0.2\n");
-	printf("\nPress ctrl+c to enter interactive Mailstation debugger\n");
+	return MS_OK;
+}
 
-	ui_init(&raw_cga_array[0], ms.lcd_dat8bit);
+int ms_run(MSHW* ms)
+{
+	// TODO: Consider removing dependency on SDL here and having
+	//     hooks for the UI code to attach to instead.
 
-	ms.z80 = z80ex_create(
-		z80ex_mread, (void*)&ms,
-		z80ex_mwrite, (void*)&ms,
-		z80ex_pread, (void*)&ms,
-		z80ex_pwrite, (void*)&ms,
-		z80ex_intread, (void*)&ms
-	);
+	int execute_counter = 0;
+	int tstate_counter = 0;
+	/* XXX: interrupt_period can change if running at different freq */
+	int interrupt_period = 187500;
+	int exitemu = 0;
+	uint32_t lasttick = SDL_GetTicks();
+	uint32_t currenttick;
+	SDL_Event event;
 
 	/* NOTE:
 	 * The z80ex library can hook in to RETI opcodes. Allowing us to exec
 	 * code here to emulate PIO or other. At this time, I don't think
 	 * there is any reason to need this.
 	 */
 
-	/* Override ctrl+c to drop to debug console */
-	sigact.sa_handler = sigint;
-	sigaction(SIGINT, &sigact, NULL);
-
 	// Display startup message
-	powerOff(&ms);
+	powerOff(ms);
 
 	lasttick = SDL_GetTicks();
 
 	while (!exitemu)
 	{
-		if (debug_console) {
-			switch (debug_prompt(&ms)) {
+		if (ms->debugger_state &= MS_DBG_ON) {
```

Comment:
&= is an assignment operator, this will do bad things here.

Reply:
Not sure how that got there...